### PR TITLE
feat[closes #38]: add workdir instruction support

### LINKED ISSUE #38: Feature request: add support for WORKDIR instruction in Vib
It would be nice to have Docker's WORKDIR instruction to specify the current working directory implement in Vib to be used with a recipe. This would be useful to be used in between modules/stages for complex applications i.e webapps, etc.

**Ref**. https://docs.docker.com/reference/dockerfile/#workdir

### PR DESCRIPTION
**Pull request by lambdaclan**
# Add `WORKDIR` instruction support

Closes #38

⚠️NOTE⚠️
---
- I am marking this PR as a draft since a lot of things require feedback from the upstream developers.
- The proposed changes introduce **breaking changes** as they alter the recipe structure meaning that existing recipes will no longer build without adopting the new style.
- I will be happy to do any necessary adjustments to be in line with the upstream developers' vision of this feature.
- The workdir related checks for each supported command are the same so maybe we can extract it to avoid duplicated code. I want to avoid refactoring/simplification until we come to an agreement of how this feature should be implemented.

## Description

This PR adds support for the [`WORKDIR`](https://docs.docker.com/reference/dockerfile/#workdir) instruction to various commands.  It also checks to ensure that unnecessary `WORKDIR` instructions are ignored when the requested working directory is already the present working directory by having a recipe level PWD. 

### Proposed Changes

- Add `WORKDIR` instruction support for all possible commands:
  - RUN, 
  - CMD
  - ENTRYPOINT 
  - COPY 
  - ADD  
- ~~Use a recipe wide PWD variable to keep track of the present working directory to avoid unnecessary (albeit harmless) instructions~~
- Restore working directory back to default path (/) after command execution
   -~~Always set the recipe's initial PWD to root (/) like default Docker~~
- Ensure that when using `singlelayer: true`  `WORKDIR` mismatches do not occur (not sure if the check is needed so TBD)
- Add support for standard COPY (not only per stage copy via --from)
- Add support for multiple adds (like copy) each with different workdir if needed
- Update recipe structure documentation
- Fix doc formatting with [prettier](https://github.com/prettier/prettier) and [mardownlint](https://github.com/DavidAnson/markdownlint)
- Fix formatting with [gofumpt](https://github.com/mvdan/gofumpt) and [goimportx](https://github.com/kesonan/goimportx)

### Local Testing

Using the following `vib.yml` recipe (based on the original [chronos-fe](https://images.vanillaos.org/#/recipe/chronos-fe) recipe):

```yml
name: Chronos Frontend Image
id: chronos-fe
stages:
  - id: build
    base: node:lts
    singlelayer: false
    labels:
      maintainer: Vanilla OS Contributors
    adds:
      - workdir: /tmp/user/files
        srcdst:
          go.mod: .
      - workdir: /tmp/otheruser/files
        srcdst:
          go.mod: .
          go.sum: .
    copy:
      - workdir: /tmp
        from: build
        paths:
          - src: /app/awesome.txt
            dst: .
          - src: /tmp/test.txt
            dst: .
      - workdir: /etc
        paths:
          - src: /app/hello.txt
            dst: .
    cmd:
      workdir: /app2
      exec:
        - python
        - prog.py
    entrypoint:
      workdir: /app
      exec:
        - npm
        - run
        - build
    runs:
      workdir: /etc/apt/apt.conf.d
      commands:
        - echo 'APT::Install-Recommends "1";' > 01norecommends
    expose:
      "6090": ""
    modules:
      - name: build-app
        type: shell
        workdir: /app
        source:
          type: git
          url: https://github.com/Vanilla-OS/chronos-frontend
          branch: main
          commit: latest
        commands:
          - mv /sources/build-app .
          - npm install
          - npm run build
repo: vanilla-os/chronos-frontend
```

```sh
vib build vim.yml
```
image

Generated Containerfile:

```dockerfile
# Stage: build
FROM node:lts AS build
WORKDIR /tmp
COPY --from=build /app/awesome.txt .
COPY --from=build /tmp/test.txt .
WORKDIR /
WORKDIR /etc
COPY /app/hello.txt .
WORKDIR /
LABEL maintainer='Vanilla OS Contributors'
WORKDIR /etc/apt/apt.conf.d
RUN echo 'APT::Install-Recommends "1";' > 01norecommends
WORKDIR /
EXPOSE 6090/
WORKDIR /tmp/user/files
ADD go.mod .
WORKDIR /
WORKDIR /tmp/otheruser/files
ADD go.mod .
ADD go.sum .
WORKDIR /
ADD includes.container /
ADD sources /sources
WORKDIR /app
RUN mv /sources/build-app . && npm install && npm run build
WORKDIR /
WORKDIR /app2
CMD ["python","prog.py"]
WORKDIR /
WORKDIR /app
ENTRYPOINT ["npm","run","build"]
WORKDIR /
```

### CODE REVIEW

**Comment by taukakao** [comment]
Please explain why you think that this feature requires breaking chages. 

Also, I think that adding state here adds unecessary complexity. I would prefer a solution that just resets the working directory to a default value (probably "/") since changing the working directory is not a costly operation.

**Comment by lambdaclan** [comment]
Hello @taukakao.

> Please explain why you think that this feature requires breaking chages.

I was trying to stay as close as possible to how an actual Dockerfile works. Each supported command can optionally specify its working directory before it executes. For each command to keep track of the working directory a new structure is needed.

Furthermore, without attaching workdir to a command state how we will know when to switch working directories? I mean YAML does not allow for duplicated keys so using workdir multiple times at the top level is not feasible. Originally I was thinking to only have a module/stage level workdir but then the workdir can only be changed once. I guess we could use other naming schemes maybe workdir+number to specify multiple workdirs at the top level but then we will need to keep track of what comes after that.

> Also, I think that adding state here adds unecessary complexity. I would prefer a solution that just resets the working directory to a default value (probably "/") since changing the working directory is not a costly operation.

Indeed, I originally had a restore flag that would optionally restore the working directory back to root after the command is executed. I am not sure if it required, but we could re-introduce it. Regarding state itself, again as mentioned above it is needed to know for which command to switch the workdir for.

I am not sure how you intend to use this feature. For example, I am not sure if it is needed for all supported commands or if you want to set a working directory multiple times in a recipe.

This is just a draft implementation of how I pictured it working, and I was a bit worried about the breaking changes but of course I am open to suggestions/feedback and will be happy to adjust as necessary. We just need to keep in mind that keeping things a bit separate will allow us to further extend Vib's feature set down the line if needed. For example having separate states for each command will allow us to add more of their options like copy has from etc.

**Review by mirkobrombin — COMMENTED** [review]
Looks good to me. I like the idea of having WORKDIR in all the available commands. Personally I do not understand where this is a breaking change. Can't we simply keep WORKDIR empty if not defined? From my understanding this would make existing recipes just works as before, correct me if I am wrong.

**Comment by lambdaclan** [comment]
Hello Mirko. Thank you very much for the feedback.

> Personally I do not understand where this is a breaking change. Can't we simply keep WORKDIR empty if not defined? From my understanding this would make existing recipes just works as before, correct me if I am wrong.

Unfortunately, that is not the case.`WORKDIR` is indeed optional but the commands' structure has also changed. For example if we look at `ENTRYPOINT` previously it was expecting just a string array whereas now it is an `EntryPoint` struct:

image

image

This means that any recipe using entrypoint for example the chronos-fe recipe will need to be updated from:

```yaml
entrypoint:
      - /bin/sh
      - '-c'
      - cd /app && npm run publish
```
to

```yaml
entrypoint:
      exec:
      - /bin/sh
      - '-c'
      - cd /app && npm run publish
```

otherwise:

image


I am wondering if there is an easy way to avoid breaking changes by using generics. Ideally what we want for  the entrypoint example is that it can either be a string[] or an EntryPoint type. This will complicate the state and the checks we need to do, but I guess it is possible. It depends on how far we want to go. I will see if I can find any other workarounds.

**Comment by mirkobrombin** [comment]
Oh, got it. I think it's fine, we already introduced breaking changes in the past, this just needs to be documented in our documentation.

**Comment by taukakao** [comment]
@lambdaclan 

I find it very important that the individual commands stay modular, so I would prefer if:

```yaml
    copy:
      - workdir: /tmp
        paths:
          - src: /app/awesome.txt
            dst: .
```

would create

```dockerfile
WORKDIR /tmp
COPY /app/awesome.txt .
WORKDIR /
```

That would cause a lot of duplicate WORKDIR instructions but I don't think that's really a problem and it would remove the need for the state and guarantees that following instructions will not be affected by the workdir.

**Comment by taukakao** [comment]
Also, if it only breaks the entrypoint then I think it would be reasonable.
Are there other commands that need to be adjusted for this?

**Comment by lambdaclan** [comment]
Thank you everyone for the feedback. 

@taukakao 
> Also, if it only breaks the entrypoint then I think it would be reasonable.
Are there other commands that need to be adjusted for this?

The new `workdir` instruction introduces a breaking change to **all** supported commands except `copy`. This is because `copy` was the only command using a custom object and could be easily extended to support additional options. This is what I mean that it might make sense to do the same for all commands to ensure we avoid reintroducing breaking changes going forwards.

### RUN

before

```yaml
runs:
  - echo "hello"
```

after:
```yaml
runs:
  comands:
    - echo "hello"
```

### CMD

before

```yaml
cmd:
  - echo "hello"
```

after:
```yaml
cmd:
  exec:
    - echo "hello"
```

### ENTRYPOINT

before

```yaml
entrypoint:
  - echo "hello"
```

after:
```yaml
entrypoint:
  exec:
    - echo "hello"
```

### ADD

before

```yaml
adds:
  myfile:/tmp
```

after:
```yaml
adds:
    srcdst:
      myfile:/tmp
    srcdst:
      myotherfile:/tmp
```

Now each of the commands above (including COPY) also takes an optional workdir to specify the working directory before each action is executed.

Can one of you please clarify the following points so that I can proceed:

- Is it OK to introduce the breaking changes above?
- Is it OK to make each workdir command automatically revert to root as requested by @taukakao?
- Is it OK to remove recipe wide PWD state? This means that we will no longer check if the workdir instruction changes to the same working directory - also depends on above point. 
  ```yaml
  runs:
    workdir: /tmp
    commands:
      - echo "hello"
  cmd:
    workdir: /tmp
    exec:
      - echo "world"
  ``` 
will generate
```dockerfile
WORKDIR /tmp
RUN echo "hello"
# WORKDIR / -> if we always revent back to root
WORKDIR /tmp
CMD [echo "world"]
# WORKDIR / -> if we always revent back to root
```
- If we are going to introduce breaking changes would you like to also update copy to behave like add? I find the current setup inconsistent because add takes a map for source/destination whereas copy take a different path object. They should both take a map or a path object.
image

- How should we handle workdir when `singlelayer = true`? Right now it seems all RUN commands get joined with commands from the modules. This can be problematic because the run command and module commands might specify a different workdir causing mismatches. Currently, if `singlelayer=true` I compare workdirs (if specified) and throw an error if they do not match.

 
Let me know when you get a chance.

Thank you!

**Comment by mirkobrombin** [comment]
> Is it OK to make each workdir command automatically revert to root as requested by @taukakao?
it is fine
> Is it OK to remove recipe wide PWD state? This means that we will no longer check if the workdir instruction changes to the same working directory - also depends on above point.
it is fine if necessary

**Comment by lambdaclan** [comment]
Updated the PR with the requested changes.  
I will be doing some more local testing while I wait for your review.
Regarding the copy/add structure mismatch I will be creating a different PR later on.

Thank you!

**Review by mirkobrombin — APPROVED** [review]
Tested with the Containerfile in the first comment. Looks good to me, nice job!
Waiting for at least another reviewer.

File: core/structs.go
Comment (taukakao) [inline core/structs.go:9]: Did you mean to change this line?

**Review by taukakao — CHANGES_REQUESTED** [review]
Just one tiny note. Otherwise it looks really good. 

Also tested building the desktop image. Just had to rebuild fsguard with the new api and change one line in the recipe and it worked.

File: core/structs.go
Comment (lambdaclan) [reply core/structs.go:9]: Hello Tau. Thank you for pointing this out. Seems like the comment was automatically removed by my autoformatting config. The issue seems to be with the `goimports-reviser` so I have a switched to a different formatter for now. I will revert the change in an upcoming commit.

**Comment by lambdaclan** [comment]
Latest commits:

- Restore comment removal (auto formatting issue)
- Update docs
- Update example and generated containerfile

Please note that CMD and ENTRYPOINT uses the exec form ["", ""] so we need to pass each command as a list which is why I used exec as a field name

```yml
entrypoint:
  exec:
    - npm run build -- wrong --> ["npm run build"]
    - npm
    - run
    -  build
["npm", "run", "build"] --> correct
```

**Review by taukakao — APPROVED** [review]
LGTM

**Review by kbdharun — APPROVED** [review]
LGTM, Thanks for your contribution. Tested the changes and checked the docs additions.

**Comment by taukakao** [comment]
Mirko already approved it, so I'm gonna merge it.

**Comment by taukakao** [comment]
Thanks for your contribution @lambdaclan

**Comment by taukakao** [comment]
(And yes, I just realized that those commits weren't squashed yet, we will decide how to deal with that)

**Pull request merged**
